### PR TITLE
adding a matrix equalization algorithm

### PR DESCRIPTION
**Pull request by meg-1**
### Describe your change:
Adding a matrix equalization algorithm, which equalizes all elements of the input vector to a common value, by making the minimal number of "updates" under the constraint of a step size (k).

This algorithm has practical use in scenarios where you need to achieve uniformity across elements of an array under specific constraints and with minimal changes. For example:
- resource allocation (this algorithm provides a way to complete the task efficiently)
- game theory (this algorithm may help in devising fair strategies using minimal moves)
- signal processing (the algorithm might be applied where signal adjustments are constrained by step size).

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Review by imSanko — APPROVED** [review]
Nice Add On!

**Comment by meg-1** [comment]
> Nice Add On!

Thank you, Sanko! @cclauss, if possible, could you please check this pr out and approve a merge? Would be much appreciated :))

**Comment by cclauss** [comment]
Please add a URL to a Wikipedia article or other source that explains what the algorithm is and why it is useful.
* https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md

**Comment by meg-1** [comment]
> Please add a URL to a Wikipedia article or other source that explains what the algorithm does and why it is useful.
> 
> * https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md

Thank you, @cclauss, just did that!

File: matrix/matrix_equalization.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import sys
+
+
+def array_equalization(vector: list[int], k: int) -> int:
```
Comment (cclauss) [inline matrix/matrix_equalization.py:4]: Self-documenting variable names, please.
```suggestion
def array_equalization(vector: list[int], step_size: int) -> int:
```

File: matrix/matrix_equalization.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    5
+    >>> array_equalization([-22, 81, -88, 71, 22, 81, 632, 81, -81, -22, 92], 2)
+    5
+    >>> array_equalization([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 5)
+    0
+    >>> array_equalization([sys.maxsize for c in range(10)], 3)
+    0
+    >>> array_equalization([22, 22, 22, 33, 33, 33], 2)
+    2
+
```
Comment (cclauss) [inline matrix/matrix_equalization.py:25]: We need tests that have the same `vector` but different values for `step-size` including -1, 0, 1.3, and sys.maxsize.

File: matrix/matrix_equalization.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import sys
```
Comment (cclauss) [inline matrix/matrix_equalization.py:1]:
```suggestion
from sys import maxsize
```

**Comment by meg-1** [comment]
> OK... Now that I know what the algorithm does, why would anyone ever want to use it?

Just added an explanation in the description of this commit:

> This algorithm has practical use in scenarios where you need to achieve uniformity across elements of an array under specific constraints and with minimal changes. For example:
> 
> resource allocation (this algorithm provides a way to complete the task efficiently)
> game theory (this algorithm may help in devising fair strategies using minimal moves)
> signal processing (the algorithm might be applied where signal adjustments are constrained by step size).

And thank you very much for the review, @cclauss , just implemented all recommendations suggested :))

**Review by cclauss — APPROVED** [review]
With https://en.wikipedia.org/wiki/Single_instruction,_multiple_data and vectorization techniques being so prevalent these daze I doubt this would ever be used but here we go...

**Pull request merged**